crates/vm/src/system/memory/merkle/trace.rs

```diff
@@ -117,33 +122,32 @@ struct TreeHelper<'a, const CHUNK: usize, F: PrimeField32> {
     memory_dimensions: MemoryDimensions,
     final_memory: &'a Equipartition<F, CHUNK>,
     touched_nodes: &'a FxHashSet<(usize, u32, u32)>,
-    trace_rows: &'a mut Vec<MemoryMerkleCols<F, CHUNK>>,
 }
 
 impl<const CHUNK: usize, F: PrimeField32> TreeHelper<'_, CHUNK, F> {
     fn recur(
-        &mut self,
+        &self,
         height: usize,
         initial_node: &MemoryNode<CHUNK, F>,
         as_label: u32,
         address_label: u32,
-        hasher: &mut impl HasherChip<CHUNK, F>,
-    ) -> MemoryNode<CHUNK, F> {
+        hasher: &(impl HasherChip<CHUNK, F> + Send + Sync),
+    ) -> (MemoryNode<CHUNK, F>, Vec<MemoryMerkleCols<F, CHUNK>>) {
```

Comment:
I think it's better if we just allocate a big 2^height array first, and then pass in pointers. can do this after we get everything working

Reply:
nevermind: (1) 2^height is too big, (2) I used `crossbeam_channel` instead